modules/transactionpool/update.go

```diff
@@ -387,3 +388,245 @@ func (tp *TransactionPool) PurgeTransactionPool() {
+		var found bool
+		var location objectLocation
+		location.setID = setID
+		for _, sci := range set[i].SiacoinInputs {
+			if oid == ObjectID(sci.ParentID) {
+				found = true
+				location.txIndex = i
+			}
+		}
+		for _, sfi := range set[i].SiafundInputs {
+			if oid == ObjectID(sfi.ParentID) {
+				found = true
+				location.txIndex = i
+			}
+		}
+		for _, fcr := range set[i].FileContractRevisions {
+			if oid == ObjectID(fcr.ParentID) {
+				found = true
+				location.txIndex = i
+			}
+		}
+		for _, sp := range set[i].StorageProofs {
+			if oid == ObjectID(sp.ParentID) {
+				found = true
+				location.txIndex = i
+			}
+		}
+		if found {
+			locations = append(locations, &location)
+		}
+	}
+
+	if len(locations) == 0 {
+		return nil, false
+	}
+	return locations, true
+}
+
+// removeTransaction removes a transaction and any transactions that spend its
+// outputs from the tpool.
+func (tp *TransactionPool) removeTransaction(txn types.Transaction, cc *modules.ConsensusChange) {
+	// removedTxnIndices is used to keep track of which need to be removed from
+	// which sets. Sets are changed (and a diff is created) once all changes
+	// have been finalized. objectLocations are used to pinpoint txn indices
+	// quickly. removedTxnIndices are kept in sorted order.
+	removedTxnIndices := make(map[TransactionSetID][]int)
+	objectLocations := make(map[ObjectID][]*objectLocation)
+
+	// removedObjects stores all object IDs of objects in transactions removed,
+	// or put in the stack to be removed.
+	removedObjects := make(map[ObjectID]struct{})
+	// removalStack is a stack of objects from removed transactions that
+	// need to be removed from other transactions.
+	var removalStack []ObjectID
+
+	// addTxnToRemovalStack adds all outputs of a transaction into the removal
+	// stack, and marks all fields (including outputs) for removal. This is to
+	// ensure that references to those objects don't remain in the
+	// transactionpool state and incur a memory-leak.
+	addTxnToRemovalStack := func(txn types.Transaction) {
+		// Add all output fields to the set of objects to remove and add
+		// them to the removal stack.
+		for i := range txn.SiacoinOutputs {
+			id := ObjectID(txn.SiacoinOutputID(uint64(i)))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+		for i := range txn.SiafundOutputs {
+			id := ObjectID(txn.SiafundOutputID(uint64(i)))
+			if _, alreadyQueuedForRemoval := removedObjects[id]; !alreadyQueuedForRemoval {
+				removedObjects[id] = struct{}{}
+				removalStack = append(removalStack, id)
+			}
+		}
+
+		// TODO: Make sure all outputs are accounted for properly in file
+		// contract related fields. Do `NewValidProofOutputs` and
+		// `NewMissedProofOutputs` fields of a FileContractRevision need to be
+		// added to the removal stack as well?
+		for i := range txn.FileContracts {
```

Comment:
This `TODO` needs to be addressed. I don't know how the `SiacoinOutputs` of a `FileContract` should be treated here.